### PR TITLE
Publish Store events to Feed

### PR DESCRIPTION
**Pull request by mrtracy**
Stores now publish intersting events to a feed. The feed is passed to NewStore
as part of the StoreContext; if no feed is passed one is generated, although it
is not currently available for subscribers.

This commit also augments MultiTestContext to optionally pass a feed to the
Stores it creates. This is used in the new test `TestMultiStoreEventFeed`, which
passes a feed to multiple stores and verifies that the correct events are
published to the feed.

Finally, this commit renames "NewStoreEvent" to "AddStoreEvent".

### CODE REVIEW

File: storage/client_event_test.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+type storeEventReader struct {
+	perStoreFeeds map[proto.StoreID][]string
+}
+
+// recordEvent records the events recieved for all stores. Each event is
```
Comment (spencerkimball) [inline storage/client_event_test.go]: s/recieved/received/

File: storage/client_event_test.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+			"EndScanRanges",
+			"SplitRange origId=1, newId=2, leader=true, origLive=934, newLive=42",
+			"SplitRange origId=2, newId=3, leader=true, origLive=42, newLive=0",
+			"MergeRange rid=2, subId=3, leader=true, live=42, subLive=0",
+		},
+	}
+	if a, e := ser.perStoreFeeds, expected; !reflect.DeepEqual(a, e) {
+		t.Errorf("event feed did not meet expected value: printing actual values to compare with above expectation:\n")
+		ser.printValues(t)
+	}
```
Comment (spencerkimball) [inline storage/client_event_test.go]: Nice test.

File: storage/feed.go [hunk 3]
```diff
@@ -22,19 +22,21 @@ import (
...
 
 // UpdateRangeEvent occurs whenever a Range is modified. This structure includes
-// the same information as NewRangeEvent, but also includes a second set of
+// the same information as AddRangeEvent, but also includes a second set of
 // MVCCStats containing the difference from the Range's previous stats. If the
 // update did not modify any statistics, this diff may be nil.
 type UpdateRangeEvent struct {
+	StoreID  proto.StoreID
 	Desc     *proto.RangeDescriptor
 	IsLeader bool
```
Comment (spencerkimball) [inline storage/feed.go]: You'll find some merge conflicts here, as I removed the IsLeader bool.

**Comment by spencerkimball** [comment]
LGTM

File: storage/client_event_test.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+				return util.Errorf("expected aa = %d, got %d", e, a)
+			}
+		}
+		return nil
+	})
+
+	// Close feed and wait for reader to receive all events.
+	feed.Close()
+	readStopper.Stop()
+	<-readStopper.IsStopped()
```
Comment (bdarnell) [inline storage/client_event_test.go]: This is redundant; `Stop` does not return until everything has stopped. `IsStopped` is only needed if you need to wait for this event from a different goroutine from the one that calls `Stop` (which is rare but we do occasionally).

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**